File: godci/src/v6/roles/contexts/accountrole.go
package contexts

import (
	"time"
	"v6/domains"
	"v6/roles/rules"
)

type AccountRole struct {
	*domains.AccountDomain
}

func (accRole *AccountRole) Deposit(accCtx *AccountDepositContext) {
	var transType string

	switch accCtx.Account.AccountType {
	case domains.ASSETACCOUNT:
		transType = domains.CREDIT
	case domains.LIABILITYACCOUNT:
		transType = domains.DEBIT
	}
	accRole.RecordTransaction(accCtx.Message, accCtx.EntryTime, accCtx.Amount, transType)
}

func (accRole *AccountRole) Withdraw(accCtx *AccountWithdrawContext) {
	var transType string

	switch accCtx.Account.AccountType {
	case domains.ASSETACCOUNT:
		transType = domains.DEBIT
	case domains.LIABILITYACCOUNT:
		transType = domains.CREDIT
	}
	accRole.RecordTransaction(accCtx.Message, accCtx.EntryTime, accCtx.Amount, transType)
}

type AccountWithdrawContext struct {
	Account       *AccountRole
	Message       string
	EntryTime     time.Time
	Amount        float64
	BusinessRules []rules.BusinessRuler
	rules.RulesRunner
}

func (accCtx *AccountWithdrawContext) Initialize(ad *domains.AccountDomain, msg string, entryTime time.Time, amount float64) {
	accCtx.Account = &AccountRole{AccountDomain: ad}
	accCtx.Message = msg
	accCtx.EntryTime = entryTime
	accCtx.Amount = amount
	accCtx.BusinessRules = []rules.BusinessRuler{AccountProtection{}}
	accCtx.RulesRunner = rules.RulesRunner{}
}

func (accCtx *AccountWithdrawContext) Execute() error {
	err := accCtx.ApplyRules(accCtx.BusinessRules, accCtx)
	if err != nil {
		return err
	}
	accCtx.Account.Withdraw(accCtx)
	return nil
}

type AccountDepositContext struct {
	Account   *AccountRole
	Message   string
	EntryTime time.Time
	Amount    float64
}

func (accCtx *AccountDepositContext) Initialize(ad *domains.AccountDomain, msg string, entryTime time.Time, amount float64) {
	accCtx.Account = &AccountRole{AccountDomain: ad}
	accCtx.Message = msg
	accCtx.EntryTime = entryTime
	accCtx.Amount = amount

}

func (accCtx *AccountDepositContext) Execute() {
	accCtx.Account.Deposit(accCtx)
}

File: jsdci/v7/roles/contexts/accountrole.js
(function() {
  "use strict";

  var Constants = require('../../domains/constants.js');
  var contexts = require('../../roles/contexts/contexts.js');
  var domains = require('../../domains/accounts.js');

  var AccountRole, AccountDepositContext, AccountWithDrawContext;

  AccountRole = {
    roleRequirements: function roleRequirements() {
      return ['accountInfo', 'entries'];
    },
    deposit:  function deposit(accountDepositContext) {
      var transType, entry;
      switch (accountDepositContext.account.accountInfo.accountType) {
        case Constants.AssetAccount:
          transType = Constants.Credit;
          break;
        case Constants.LiabilityAccount:
          transType = Constants.Debit;
          break;
      }
      entry = new domains.EntryItem(accountDepositContext.account.accountInfo.accountID,
                                      accountDepositContext.entryTime,
                                      accountDepositContext.message,
                                      accountDepositContext.amount, transType);
      accountDepositContext.account.entries[accountDepositContext.account.entries.length] = entry;
    },
    withdraw: function withdraw(accountWithDrawContext) {
      var transType, entry;
      switch (accountWithDrawContext.account.accountInfo.accountType) {
        case Constants.AssetAccount:
          transType = Constants.Debit;
          break;
        case Constants.LiabilityAccount:
          transType = Constants.Credit;
          break;
      }
      entry = new domains.EntryItem(accountWithDrawContext.account.accountInfo.accountID,
                                    accountWithDrawContext.entryTime,
                                    accountWithDrawContext.message,
                                    accountWithDrawContext.amount, transType);
    accountWithDrawContext.account.entries[accountWithDrawContext.account.entries.length] = entry;
    }
  };

  AccountDepositContext = function AccountDepositContext(accountDomain, message, entryTime, amount) {
    var self = this instanceof AccountDepositContext ? this : Object.create(AccountDepositContext.prototype);
    self.account = accountDomain;
    self.message = message;
    self.entryTime = entryTime;
    self.amount = amount;
    self.assignRole(AccountRole, self.account, this);
    return self;
  };

  AccountDepositContext.prototype = Object.create(contexts.Context.prototype);

  AccountDepositContext.prototype.execute = function execute() {
    this.account.deposit(this);
    this.removeRole(AccountRole, this.account);
  };

  AccountWithDrawContext = function AccountWithDrawContext(accountDomain, message, entryTime, amount) {
    var self = this instanceof AccountWithDrawContext ? this : Object.create(AccountWithDrawContext.prototype);
    self.account = accountDomain;
    self.message = message;
    self.entryTime = entryTime;
    self.amount = amount;
    self.assignRole(AccountRole, self.account, this);
    return self;
  };

  AccountWithDrawContext.prototype = Object.create(contexts.Context.prototype);

  AccountWithDrawContext.prototype.execute = function execute() {
    this.account.withdraw(this);
    this.removeRole(AccountRole, this.account);
  };

  module.exports = { AccountRole: AccountRole, AccountDepositContext: AccountDepositContext,
    AccountWithDrawContext: AccountWithDrawContext
  };
})();
